Restructure automatic checks docs sections (#573)
canonical/sphinx-docs-starter-pack: docs/how-to/run-documentation-checks.rst

.. _inclusive_lang_check:

Inclusive language check
------------------------

The Starter Pack checks for inclusive language with a custom set of `Vale`_ rules. To check for inclusive language violations, run:

.. code-block:: bash

    make woke

If you need to exempt single words or sections, follow the process in :ref:`vale_exemptions`. 

.. _link_check:

Link check
----------

The Starter Pack checks any links in your documentation with the Sphinx ``linkcheck`` builder. To validate the links, run:

.. code-block:: bash

    make linkcheck

Configure
~~~~~~~~~

If you have links in the documentation that you don't want to check, add them to the ``linkcheck_ignore`` list in the ``conf.py`` file.

.. _markdown_lint_check:

Markdown lint check
-------------------

The Starter Pack checks standards and consistency in Markdown files with the Markdown lint check. To check your Markdown files, run:

.. code-block:: bash

    make lint-md

> Configure

Reviewer: Inconsistent use of a "Configure" subsection in some of these sections, but not in others that include the same type of information without a "Configure" subsection. Either remove the "Configure" subsection from all (I prefer this) or add it to all.

Author: Do you have specific examples? Outside of the vale exemption statement, which I'd say is separate from configuration, I can only find some of the additional style guide linting details but I wouldn't say those are really configuration either.

Reviewer: What I mean is, if sentences like this:

```
If you need to exempt single words or sections, follow the process in [Vale exemptions](https://canonical-starter-pack--573.com.readthedocs.build/573/how-to/run-documentation-checks/#vale-exemptions).
```

don't need to be put under a "Configure" subsection, then why have a "Configure" subsection for sentences like this?:

```
### Configure

If you have links in the documentation that you don’t want to check, add them to the linkcheck_ignore list in the conf.py file.
```

The way it makes it look in the table of contents is that only those checks can be configured:
<img width="225" height="329" alt="image" src="https://github.com/user-attachments/assets/95878317-4354-420a-b4d4-7465a2a36141" />

Author: That's true - I'd been thinking about it as aspects that are done inline (like the exemptions) versus what needs to be done ahead of time (in a different file) but Vale uses the custom_wordlist file so that argument doesn't even hold there. I agree only having the Configure heading on some makes it look like the others don't have options. I'll update to all having the configure header for now and it can be revisited next cycle.

.. code-block:: bash

    make spelling

If you need to exempt single words or sections, follow the process in :ref:`vale_exemptions`. 

.. _style_guide_linting:

Style guide linting
-------------------

The Starter Pack runs the `Vale`_ documentation linter configured with the rules of `style guide <https://github.com/canonical/documentation-style-guide>`__ To check your documentation with Vale, run:

.. code-block:: bash

    make vale

Vale can run against individual files, directories, or globs. To check a specific target, run:

.. code-block:: bash

    make vale TARGET=example.file
    make vale TARGET=example-directory

If you run the linter against a directory, its subdirectories ware also linted.

You can use wildcards to lint globs of files. For example, to run against all Markdown files within a directory, you'd run:

.. code-block:: bash

    make vale TARGET=*.md

If you need to exempt single words or sections, follow the process in :ref:`vale_exemptions`. 

.. _vale_exemptions:
.. _reference-automatic-checks-spelling-vale-ignore:

> Vale exemptions

Reviewer: "Vale exemptions" as a section does not match the other sections at the same heading level. I suggest putting all the other sections under heading called "Checks" then "Vale exemptions" as a section at the same level as "Checks".

As in:
- Checks
  - Accessibility
  - Inclusive language
  - Etc
- Vale exemptions

Rather than:
- Accessibility
- Inclusive languge
- Etc
- Vale exemptions

Author: I've been trying to keep the levels of headers minimal since it was getting busy very quickly. If it's ok with you, let's deffer this one to next cycle's work. I agree though that it doesn't fit with the actual checks.

.. _run-documentation-checks :

Run documentation checks
========================

The starter pack comes with several tests and checks that you can (and should!) run on your documentation before committing and pushing changes:

- :ref:`accessibility_check`
- :ref:`inclusive_lang_check`
- :ref:`link_check`
- :ref:`markdown_lint_check`
- :ref:`spelling_check`
- :ref:`style_guide_linting`
- :ref:`vale_exemptions`

.. _accessibility_check:

> Accessibility check

Reviewer 2: Suggested change
      
    
    
      
          
            
            Accessibility check
          
          
            
            Check accessibility
          
      
    
    
  

I prefer when how-to sections are imperative, but it's not a hill I'd die on. This would obviously need to be uniform across the doc.

Nouns as section titles always makes me think I'm looking at an explanation or reference page.

-------------------

The Starter Pack checks the accessibility of the documentation with `Pa11y`_. It's configured to check for Level AA conformity to the `Web Content Accessibility Guidelines (WCAG) 2.2`_.

> Note that this check is only available locally; it is not part of the Starter Pack's :ref:`github-workflows`.

Reviewer 2: ```suggestion
This check is only available locally; it is not part of the Starter Pack's :ref:`github-workflows`.
```

> **Prerequisites:** ``Pa11y`` must be installed through ``npm``. Install ``npm`` with the appropriate method for your operating system; follow the `npm docs installation guide <https://docs.npmjs.com/downloading-and-installing-node-js-and-npm>`_ for details. 

Reviewer 2: Optional:
```suggestion
To run the check, you must have ``Pa11y`` installed.

.. code-block:: bash

    npm install pa11y

If you don't have npm installed, refer to the `installation guide <https://docs.npmjs.com/downloading-and-installing-node-js-and-npm>`__. 
```

Like 'make', npm is one of those things that's become so natural in software that I don't think it warrants monospacing.

Also, we should generally use two underscores for external links. Single-underscore links create named references that can be reused, while double-underscore links remain anonymous.

To check the accessibility of the documentation, run the following command from within your ``/docs`` directory:

.. code-block:: bash

>     make pa11y

Reviewer 2: Throughout this page, there are code blocks that use two, three, and four spaces for indentation.

I find four spaces to be the most visually accessible, but I'm not picky as long as they're consistent.

Author: Ah thanks for catching that! I use tab pretty much exclusively but I think when I was cutting and pasting some of the formatting got messed up.

Configure
~~~~~~~~~

> The ``pa11y.json`` file in the ``.sphinx`` directory provides basic defaults expected to be used for all teams; refrain from removing any unless absolutely necessary. To include additional checks, browse the available settings and options on ``Pa11y``'s `README <Pa11y readme_>`_ on GitHub.

Reviewer 2: Optional:
```suggestion
The ``pa11y.json`` file in the ``.sphinx`` directory provides basic defaults expected to be used for all teams; refrain from removing any unless absolutely necessary. Additional settings and options are detailed in the ``Pa11y`` `README <https://github.com/pa11y/pa11y#command-line-configuration>`__ on GitHub.
```
This is pedantic, but the sentence attached to this link makes it seem like the action of browing the README will add additional checks to the reader's project.

Also, this reST link scares me :o

```
``Pa11y``'s `README <Pa11y readme_>`_
```

I think I'd rather have the source be readable here than reuse an existing link. Since this isn't a super common link, I actually think having to track down the original declaration makes this _harder_ to maintain.

.. _inclusive_lang_check:

Inclusive language check
------------------------

The Starter Pack checks for inclusive language with a custom set of `Vale`_ rules. To check for inclusive language violations, run:

.. code-block:: bash

    make woke

> For options on exempting single words or sections, follow the process in the :ref:`vale_exemptions` section. 

Reviewer 2: Optional:
```suggestion
If you need to exempt single words or sections, follow the process in :ref:`vale_exemptions`. 
```
This seems nitpicky, but I think there's real value in the inclusion of 'need to'. It makes it clear that this isn't something to be done when it's convenient.

.. _link_check:

Link check
----------

The Starter Pack checks any links in your documentation with the Sphinx ``linkcheck`` builder. To validate the links, run:

.. code-block:: bash

   make linkcheck

Configure
~~~~~~~~~

> If you have links in the documentation that you don't want to check, you can add them to the ``linkcheck_ignore`` variable in the ``conf.py`` file.

Reviewer 2: Optional:
```suggestion
If you have links in the documentation that you don't want to check, add them to the ``linkcheck_ignore`` list in the ``conf.py`` file.
```
No need for the passivity here imo

.. _markdown_lint_check:

Markdown lint check
-------------------

The Starter Pack checks standards and consistency in Markdown files with the Markdown lint check. To check your Markdown files, run:

.. code-block:: bash

   make lint-md

Configure
~~~~~~~~~

> You can update the linting rules to be enforced in the ``.sphinx/.pymarkdown.json`` file. Refer to `the pymarkdown rules documentation <https://pymarkdown.readthedocs.io/en/latest/rules/>`_ for all the available rules.

Reviewer 2: Optional:
```suggestion
You can update the linting rules in the ``.sphinx/.pymarkdown.json`` file. Refer to `the pymarkdown rules documentation <https://pymarkdown.readthedocs.io/en/latest/rules/>`_ for all the available rules.
```
This redundancy caught my attention.

.. _spelling_check:

Spelling check
--------------

> The Starter Pack uses `Vale`_ to check the spelling in your documentation. It ignores code (both code blocks and inline code) and URLs (but it does check the link text). To ensure there are no spelling errors in your documentation, run:

Reviewer 2: Optional:
```suggestion
The Starter Pack uses `Vale`_ to check the spelling in your documentation. It ignores inline code, code blocks, and URLs, but it does check the link text. To ensure there are no spelling errors in your documentation, run:
```
I just don't like parenthetical clauses. 🤷

.. code-block:: bash

  make spelling

> For options on exempting single words or sections, follow the process in the :ref:`vale_exemptions` section. 

Reviewer 2: Optional: same as https://github.com/canonical/sphinx-docs-starter-pack/pull/573/changes#r3103457255

.. _style_guide_linting:

Style guide linting
-------------------

> The Starter Pack includes a method to run the `Vale`_ documentation linter configured with `the Vale rules for the current style guide <Vale rules_>`_ To check your documentation with, run:

Reviewer 2: Optional:
```suggestion
The Starter Pack runs the `Vale`_ documentation linter configured with the rules of the `style guide <https://github.com/canonical/documentation-style-guide>`__ To check your documentation with, run:
```

Required:

```diff
- To check your documentation with, run:
+ To check your documentation with Vale, run:
```

.. code-block:: bash

   make vale

Vale can run against individual files, directories, or globs. To check a specific target, run:

.. code-block:: bash

    make vale TARGET=example.file
    make vale TARGET=example-directory

> Note that running the linter against a directory will also run against its subdirectories.

Reviewer 2: Optional:
```suggestion
If you run the linter against a directory, its subdirectories are also linted.
```

> You can use wildcards to run against all files matching a string, or an extension. For example, to run against all :code:`.md` files within a directory:

Reviewer 2: Optional:
```suggestion
You can use wildcards to lint globs of files. For example, to run against all Markdown files within a directory, you'd run:
```

.. code-block:: bash

    make vale TARGET=*.md

> For options on exempting single words or sections, follow the process in the :ref:`vale_exemptions` section. 

Reviewer 2: Optional: same as https://github.com/canonical/sphinx-docs-starter-pack/pull/573/changes#r3103457255

.. _vale_exemptions:
.. _reference-automatic-checks-spelling-vale-ignore:

Vale exemptions

> ---------------

Reviewer 2: Optional:

For the subsections, I think we can convey the same meaning with:

* Words
  * Since the sections for "single instance" and "every instance" are adjacent and fit on a single page fold, I'd merge them.
* Paragraphs
* Directives

> The :ref:`inclusive_lang_check`, :ref:`spelling_check`, and :ref:`style_guide_linting` all use `Vale`_ and share a common ``.custom_wordlist.txt`` file and follow the same process to add exemptions. The Vale repository `includes a common list of words <https://github.com/canonical/documentation-style-guide/blob/main/styles/config/vocabularies/Canonical/accept.txt>`_ that will be excluded from the checks. 

Reviewer 2: ```suggestion
The :ref:`inclusive_lang_check`, :ref:`spelling_check`, and :ref:`style_guide_linting` all use `Vale`_. These checks share a common syntax for exemptions and pull from the same ``.custom_wordlist.txt`` file and

The style guide repository `includes a common list of words <https://github.com/canonical/documentation-style-guide/blob/main/styles/config/vocabularies/Canonical/accept.txt>`_ that is excluded from the checks. 
```

> Note that Vale will check the displayed text of a link, not the URL of a link. If you wish to use a link that contains non-inclusive language or incorrect spelling, use appropriate link text with the syntax appropriate for your source file. 

Reviewer 2: This is the second time this is mentioned, and it now appears before the relevant syntax. Drop?

A single instance of a word
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> To ignore only a single instance of a word, you can use the ``:vale-ignore:`` role (for spelling or style guide) or ``:woke-ignore:`` role (for inclusive language) and ensure your ``conf.py`` file contains the appropriate class association in the ``rst_prolog``:

Reviewer 2: Optional:
```suggestion

To ignore a single instance of a word, wrap it in the ``:vale-ignore:`` role. Ensure that your ``conf.py`` file contains the following ``rst_prolog``:
```
I'm fairly confident that the woke-ignore role is only needed in legacy versions at this point. If I'm wrong, then it'll only matter until I redesign the workflows in a few weeks.

.. code-block:: text

  rst_prolog = """
  .. role:: vale-ignore
      :class: vale-ignore
  .. role:: woke-ignore
      :class: woke-ignore
  """

Every instance of a word
~~~~~~~~~~~~~~~~~~~~~~~~

> To ignore a word across your documentation, add the word to the custom exceptions for your project in the ``.custom_wordlist.txt`` file.

Reviewer 2: Optional:
```suggestion
~~~~~~~~~~~~~~~~~~~~~~~~

To ignore a word across your documentation, add it to your project's ``.custom_wordlist.txt`` file.
```

> .. note::

Reviewer 2: Optional:
```suggestion
.. admonition:: Capitalization
    :class: note
```
Same appearance with a more descriptive title

>    Entries in ``.custom-wordlist`` are case-sensitive only when a capitalised word is used. For instance:

Reviewer 2: Optional:
```suggestion
   Entries in ``.custom-wordlist`` are case-sensitive only when the word is capitalized. For instance:
```

   - Adding ``kustom`` will cause all instances of ``Kustom`` and ``kustom`` to be ignored.
   - Adding ``Kustom`` will cause only instances of ``Kustom`` to be ignored.

A section of text

> ~~~~~~~~~~~~~~~~~

Reviewer 2: ```suggestion
~~~~~~~~~~~~~~~~~

```

> To disable Vale for a specific section of text, specific markup can be used:

Reviewer 2: Optional:
```suggestion
To disable Vale for a section of text, use the following markup:
```

.. tab-set:: 

    .. tab-item:: Markdown

        .. code-block:: text

            <!-- vale off -->

            This text will be ignored.

            <!-- vale on -->
    
    .. tab-item:: reST

        .. code-block:: text

            .. vale off

            This text will be ignored.

            .. vale on

A specific directive block
~~~~~~~~~~~~~~~~~~~~~~~~~~

> To disable Vale linting for a specific directive, such as a code block or note, you can apply a class to the section:

Reviewer 2: ```suggestion
To disable Vale linting for a specific directive, such as a code block or admonition, apply the `vale-ignore` class to the section:
```

.. tab-set:: 

    .. tab-item:: Markdown

        .. code-block:: text

            ````{class} vale-ignore
            ```{code-block}

            This content will be ignored by Vale.
            ```
            ````
    
    .. tab-item:: reST

        .. code-block:: text

            .. class:: vale-ignore
                .. code-block::

                    This content will be ignored by Vale.

> Note that the class process should not be necessary for Markdown, as Vale has an expanded scope for ignoring Markdown content by default. Also, the `.. class::` directive for reST does not need to encapsulate content, it applies to the next logical block (which can be another directive or even a paragraph of content). The class option should only be used when the other options are not suitable.

Reviewer 2: Optional:
```suggestion
The class process isn't necessary for Markdown, as Vale has an expanded scope for ignoring Markdown content by default. Also, the `.. class::` directive for reST does not need to encapsulate content, it applies to the next logical block (which can be another directive or even a paragraph of content). The class option should only be used when other options aren't suitable.
```

.. important::

    The checks might still flag some terms that contain hyphens or spaces.

>     For example, "Juju 3" was unable to be ignored by this method, and `needed to be added to the a specific exception within a rule <https://github.com/canonical/documentation-style-guide/blob/a6f530b07d774bee67dd79d146ae5bbedc9ddef1/styles/Canonical/013-Spell-out-numbers-below-10.yml#L15>`_.

Reviewer 2: ```suggestion
.. note::

    The checks may flag terms that contain hyphens or spaces.

    For example, "Juju 3" wasn't ignored by default and `needed to be added to be excepted with an upstream rule <https://github.com/canonical/documentation-style-guide/blob/a6f530b07d774bee67dd79d146ae5bbedc9ddef1/styles/Canonical/013-Spell-out-numbers-below-10.yml#L15>`_.
```
The 'important' class seems overkill for this.

.. _run-documentation-checks :

Run documentation checks
========================

The starter pack comes with several tests and checks that you can (and should!) run on your documentation before committing and pushing changes:

- :ref:`accessibility_check`
- :ref:`inclusive_lang_check`
- :ref:`link_check`
- :ref:`markdown_lint_check`
- :ref:`spelling_check`
- :ref:`style_guide_linting`

.. _accessibility_check:

Accessibility check
-------------------

The Starter Pack checks the accessibility of the documentation with `Pa11y`_.

It's configured to check for Level AA conformity to the `Web Content Accessibility Guidelines (WCAG) 2.2`_.

.. note::

   This check is only available locally; it is not part of the CI workflows.

Prerequisites
~~~~~~~~~~~~~

``Pa11y`` must be installed through ``npm``. Install ``npm`` with the appropriate method for your operating system: 

* Your preferred package manager 
* By following the `node version manager installation process <https://docs.npmjs.com/downloading-and-installing-node-js-and-npm#using-a-node-version-manager-to-install-nodejs-and-npm>`_
* For Debian and Ubuntu Linux distributions, the ``sudo apt install npm`` command

Run
~~~

> To check the accessibility of the documentation, run the following command from within your ``/docs`` folder.

Reviewer 2: Optional:
```suggestion
To check the accessibility of the documentation, run the following command in your ``/docs`` directory.
```
* "from within" -> "in": my preference for concision
* "folder" -> "directory": This is just a pet peeve of mine. "folder" make me think of Windows >:(

.. code-block:: bash

    make pa11y

Configure
~~~~~~~~~

> The :file:`pa11y.json` file in the :file:`.sphinx` folder provides basic defaults.

Reviewer 2: Optional:
```suggestion
The ``pa11y.json`` file in the ``.sphinx`` folder provides basic defaults.
```
IIRC, there used to be something in the style guide that said to use these roles, but I think they just add noise to the source since they just monospace the text anyways.

To browse the available settings and options, see ``Pa11y``'s `README <Pa11y readme_>`_ on GitHub.

.. _inclusive_lang_check:

Inclusive language check
------------------------

The inclusive language check uses `Vale`_ to check for violations of inclusive language guidelines.

Run
~~~

> Run the following command from within your ``/docs`` folder::

Reviewer 2: ```suggestion
Run the following command from within your ``/docs`` folder:
```
This is messing up how the next code block renders.

.. code-block:: bash

    make woke

> Exempt a word in a single instance

Reviewer 2: Optional:
```suggestion
Exempt an instance of a word
```

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

To exempt an individual word, wrap it in the ``woke-ignore`` role::

   :woke-ignore:`<SOME_WORD>`

For instance::

>    This is your text. The word in question is here: :woke-ignore:`whitelist`.

Reviewer 2: ```suggestion
For instance:

.. code-block:: text

    This is your text. The word in question is here: :woke-ignore:`whitelist`.
```

.. note::

   Vale will lint the displayed text of a link, not the URL of a link. If you
   wish to use a link that contains non-inclusive language, use appropriate link

>    text with the syntax appropriate for your source file. 

Reviewer 2: Optional:
```suggestion
   Vale lints the display text of a link, but not its URL. To ignore
   link text that uses non-inclusive language, wrap it in the role like
   you would with any other text.
```

.. note::

   Entries in ``.custom-wordlist`` are case-sensitive only when a capitalised word is used. For instance:

   - Adding ``kustom`` will cause all instances of ``Kustom`` and ``kustom`` to be ignored.
   - Adding ``Kustom`` will cause only instances of ``Kustom`` to be ignored.

Exclude multiple lines from a file
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Vale can be switched on and off within a file using syntax appropriate to that
format.

.. _link_check:

Link check
----------

The link check uses Sphinx to access the links in the documentation output and validate whether they are working.

Run
~~~

Run the following command from within your ``/docs`` folder.

Validate links within the documentation::

   make linkcheck

Configure
~~~~~~~~~

If you have links in the documentation that you don't want to check, you can add them to the ``linkcheck_ignore`` variable in the :file:`conf.py` file.

.. _markdown_lint_check:

Markdown lint check
-------------------

The Markdown lint check is used to enforce standards and consistency in Markdown files.

Run
~~~

Run the following command from within your ``/docs`` folder to lint your Markdown files::

>    make lint-md

Reviewer 2: ```suggestion
Run the following command from within your ``/docs`` folder to lint your Markdown files:

.. code-block:: bash

   make lint-md
```

Configure
~~~~~~~~~

You can update the linting rules to enforce in the :file:`.sphinx/.pymarkdown.json` file. Refer to `the pymarkdown rules documentation <https://pymarkdown.readthedocs.io/en/latest/rules/>`_ for all the available rules.

.. _spelling_check:

Spelling check
--------------

The spelling check uses ``vale`` to check the spelling in your documentation.
It ignores code (both code blocks and inline code) and URLs (but it does check the link text).

Run
~~~

Run the following commands from within your ``/docs`` folder.

Ensure there are no spelling errors in the documentation::

>   make spelling

Reviewer 2: Can you combine these sentences like in the other checks? We should fix the code block, too.

> Configure

Reviewer 2: Optional:

The "Configure" section contains instructions on global exclusion, and "Exclude specific terms" is the same for local. When I read the latter title, I'd personally expect to find global exclusion instructions.

I think we have one section called something like "Exclude words" that provides instructions for both. Ignore my specific wording, but I mean something like:

> Exclude words
> ---------------------
>
> To ignore a word project-wide, ...
> 
> To ignore only a single instance of the word, ...

Like the "run" thing, this is another area where the doc structure feels over-architected to me.

~~~~~~~~~

The Vale repository `includes a common list of words <https://github.com/canonical/documentation-style-guide/blob/main/styles/config/vocabularies/Canonical/accept.txt>`_ that will be excluded from the check.
To add custom exceptions for your project, add them to the :file:`.custom_wordlist.txt` file.

.. _reference-automatic-checks-spelling-vale-ignore:

Exclude specific terms
~~~~~~~~~~~~~~~~~~~~~~

> Sometimes, you need to use a term in a specific context that should usually fail the spelling check.

Reviewer 2: Optional:
```suggestion
Sometimes, you need to use a term that fails the spelling check.
```

.. _run-documentation-checks :

Run documentation checks

> ====================

Reviewer 2: Required tedium:
```suggestion
Run documentation checks
========================
```

The starter pack comes with several tests and checks that you can (and should!) run on your documentation before committing and pushing changes.

> The avaiable checks are:

Reviewer 3: ```suggestion
The starter pack comes with several tests and checks that you can (and should!) run on your documentation before committing and pushing changes:
```
(non-blocking suggestion)

- :ref:`accessibility_check`
- :ref:`inclusive_lang_check`
- :ref:`link_check`
- :ref:`markdown_lint_check`
- :ref:`spelling_check`

> - :ref:`style_guide_linting`

Reviewer 2: Optional: Doubling down on @a-velasco's comment ([this one](https://github.com/canonical/sphinx-docs-starter-pack/pull/573/changes#r3028916240)), I've always found this pattern to be redundant when it mirrors the local navigation this closely.

Not super opinionated on this—just something to think about.

Author: I added that partially because the side nav can be hidden if the page view is narrow enough (no idea what it does on mobile) and even with the side nav it takes a minute to parse it and get an idea of what the full list of checks are.

Reviewer 3: I generally agree with @jahn-junior and prefer to omit lists of contents that are already shown in the navigation.

In this page specifically, though, for some reason don't mind it. I think it doesn't feel the same as other guides because these sections are not sequential and are rather independent from one another. This list at the beginning feels like a useful snippet of a reference/lookup table to get some context about all the checks available.

Also not super opinionated, so I'd propose leaving it like this for now and discussing it more offline - there are folks working on standardizing these kind of structural patterns across the SP docs, so it might be an interesting conversation in that scope.

.. _accessibility_check:

Accessibility check
-------------------

> The accessibility check uses `Pa11y`_ to check for accessibility issues in the documentation output.

Reviewer 2: Optional:
```suggestion
The Starter Pack checks the accessibility of your documentation with `Pa11y`_.
```
> The accessibility check uses Pa11y

In my mind, the accessibility check _is_ Pa11y. I also just think "use" is a very nondescript verb, so I like to avoid it. This rewording could be propagated to the other checks as well.

Author: Is it really the Starter Pack that's checking? That's my one hang-up with this wording, but I'm not sure of a decent alternative.

Reviewer 2: In a transitive kinda way, yes. Since the Starter Pack configures all of the checks and abstracts away the actual commands, I think the wording is reasonable, especially from a user's perspective. If we were giving instructions on running the actual commands directly, I'd say otherwise.

If I'm being honest, this was just the best way I could think of to clean up the wording. Not super picky about the specifics.

> It is configured to use the `Web Content Accessibility Guidelines (WCAG) 2.2`_, requiring `Level AA conformance`_.

Reviewer 2: ```suggestion
It's configured to check for Level AA conformity to the `Web Content Accessibility Guidelines (WCAG) 2.2`_.
```

.. note::

>    This check is only available locally.

Reviewer 3: ```suggestion
   This check is only available locally; it is not part of the CI workflows.
```

(non-blocking suggestion - just one of those times where I feel a little bit of redundancy can be clarifying)

Prerequisites
~~~~~~~~~~~~~

> ``Pa11y`` must be installed through ``npm``. Install ``npm`` using the appropriate method for your operating system through one of the following methods: 

Reviewer 3: ```suggestion
``Pa11y`` must be installed through ``npm``. Install ``npm`` using the appropriate method for your operating system: 
```
(non-blocking suggestion)

Reviewer 2: Going to piggyback on this one:
```suggestion
``Pa11y`` must be installed through ``npm``. Install ``npm`` with the appropriate method for your operating system: 
```
Continuing my crusade against the word "use"

* Your preferred package manager 
* By following the `node version manager installation process <https://docs.npmjs.com/downloading-and-installing-node-js-and-npm#using-a-node-version-manager-to-install-nodejs-and-npm>`_

> * For Debian and Ubuntu Linux distributions, the ``sudo apt install npm`` command

Reviewer 3: ```suggestion
* For Debian and Ubuntu Linux distributions, the ``sudo apt install npm`` command
* By following the `node version manager installation process <https://docs.npmjs.com/downloading-and-installing-node-js-and-npm#using-a-node-version-manager-to-install-nodejs-and-npm>`_
* Your preferred package manager 
```
(non-blocking suggestion)

Reviewer 2: I wonder if we could be authoritative here and just say that they should use the version manager. Unless things have changed since the last time I installed NPM, this is the recommended method.

Author: These specific instructions have gone through multiple iterations and have probably accumulated some unnecessary excess. Is there any reason why the node version manager process wouldn't work - OS, etc? If not, I'm happy to update this to just refer to the preferred method.

Reviewer 3: From what I can see, the npm docs already go through a bunch of different options for different operating systems, so I think it makes sense to just refer to that link. 

I'd even remove the anchor and leave the link as https://docs.npmjs.com/downloading-and-installing-node-js-and-npm

(it's got some useful context in the introduction for people who aren't very familiar with the Node/JS world)

Once ``npm`` is installed, install ``Pa11y`` by running this command from within your documentation folder.

.. code-block:: bash

>    make pa11y-install

Reviewer 2: It should do this automatically if users don't have Pa11y installed when they run `make pa11y`.

Run
~~~

Run the following command from within your documentation folder.

Look for accessibility issues in rendered documentation::

>    make pa11y

Reviewer 2: Optional:
```suggestion
To check the accessibility of your documentation, run:

.. code-block:: bash

    make pa11y
```
Once we set the precedent that these commands are run from the root of the docs, we probably don't need to mention it afterward.

Regarding the code block, I think this should be propagated throughout the rest of the document. Code-formatting with indentation isn't very accessible (how topical :P).

Configure
~~~~~~~~~

The :file:`pa11y.json` file in the :file:`.sphinx` folder provides basic defaults.

> To browse the available settings and options, see ``Pa11y``'s `README <Pa11y readme_>`_ on GitHub.

Reviewer 2: I'm going to say that we shouldn't be highlighting that this is configurable. Ideally, everybody should be running the same checks. Very few projects run this check, but it's still an important principle.

Author: Maybe we could elaborate that this would only be if there's something specific that someone wants to *add* and no one should remove the checks provided to everyone?

Reviewer 3: Agree with @AshleyCliff here.

I understand the importance of having everyone running the same checks, but there will _always_ be exceptions and unexpected needs.I'd rather be clear and opinionated about our recommendations, but still provide clear guidance on the right way of customizing rather than omitting that information entirely.

Reviewer 2: My hang-up is that we're testing against an external standard here. If someone ignores a rule, the docs are no longer compliant. However, with this being an optional check already, I won't make a fuss.

I would like if we help users decide _when_ to do this like Ashley suggested, though.

.. _inclusive_lang_check:

Inclusive language check
------------------------

> The inclusive language check uses `Vale`_ to check for violations of inclusive language guidelines.

Reviewer 2: Optional:
```suggestion
The Starter Pack checks for inclusive language with a custom set of `Vale`_ rules.
```
Similar to my suggestion for the accessibility check. This cuts down to one instance of "inclusive language" in the sentence. Feel free to adjust this to your liking.

Run
~~~

Run the following command from within your documentation folder::

   make woke

Configure
~~~~~~~~~

> By default, the inclusive language check is applied to Markdown and |RST| files located in the documentation folder (usually :file:`docs/`).

Reviewer 2: This section doesn't contain an action. I think we could safely drop this.

Exemptions
~~~~~~~~~~

Sometimes, you might need to use some non-inclusive words.

> In such cases, you may exclude them from the check.

Reviewer 2: Optional: I think this bit is unnecessary. I don't think we'd be losing any information by bumping the following sections up a level, but that may be a matter of taste.

I think we could probably make the whole ToC shallower, but I'll wait until I can see it rendered before critiquing anything else.

Author: Agreed, this header isn't necessary since the sub-headings are very specific. (The overall TOC can't get too much more shallow unless we drop the Run/Config/etc. as sub-headings but I'm open to options - I don't love how it looks in the side nav currently)

Exempt a word in a single instance
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> To exempt an individual word, give the word the ``woke-ignore`` role::

Reviewer 2: Optional:
```suggestion
To exempt an individual word, wrap it in the ``woke-ignore`` role::
```

Reviewer 2: Also, a question for @a-velasco:

Is this the role that we still needed because woke is installed in the composite GitHub action? I seem to recall you raising this earlier in the cycle.

Reviewer 3: Yeah I think so - the one that gets called by the composite `documentation-workflows` action is [still based on `woke`](https://github.com/canonical/documentation-workflows/blob/main/inclusive-language/inclusive_language.py) as far as I can tell.

Reviewer 2: I'm curious why that is. The workflow doesn't _invoke_ it at any point (unless the project's Makefile still does), but maybe it's presence affects something. :shrug: 

It at least won't be a concern after the workflow epic next cycle.

   :woke-ignore:`<SOME_WORD>`

For instance::

>    This is your text. The word in question is here: :woke-ignore:`whitelist`.

Reviewer 2: Optional: I'd prefer if we came up with a different example. This one shows how to ignore a word that teams really shouldn't be ignoring in their docs.

I'm not sure if we have any disallowed words that reside in more of a grey area, but it's something to think about.

Author: Agreed a different example would be best; I'll try to come up with one (suggestions welcome).

Reviewer 4: A common one I have to exempt frequently is "master" - most upstreams have moved away from whitelist, but some still use "master" branches and haven't yet made the switch to "main" or similar.

.. note::

   Vale will lint the displayed text of a link, not the URL of a link. If you
   wish to use a link that contains non-inclusive language, use appropriate link
   text with the syntax appropriate for your source file. 

Exempt a word globally
^^^^^^^^^^^^^^^^^^^^^^

Vale will ignore any word listed in the ``.custom_wordlist.txt`` file.

> To exempt a word, add it to this file globally.

Reviewer 2: Optional:
```suggestion
Vale ignores any words in the ``.custom_wordlist.txt`` file.
To exempt a word, add it to this file.
```
I don't like future tense >:(

Reviewer 3: I kinda like that it makes things sound vaguely prophetic.

(+1 to the suggestion though)

.. note::

>    Entries in ``.custom-wordlist`` are case-sensitive only when a capitalised word is used. For instance:

Reviewer 3: My mind just got blown. TIL.

   - Adding ``kustom`` will cause all instances of ``Kustom`` and ``kustom`` to be ignored.
   - Adding ``Kustom`` will cause only instances of ``Kustom`` to be ignored.

Exclude multiple lines from a file
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

Vale can be switched on and off within a file using syntax appropriate to that

> format.

Reviewer 2: Required: What's the syntax? It's not clear to me at this point in the doc.

Author: I've been hoping someone would chime in to clarify this one.. I have no idea what the implementation actually looks like from this statement.

Author: Whoops that was my bad - there was a code block that I was supposed to have copied over with that text but missed. It's been fixed.

.. _link_check:

Link check
----------

> The link check uses Sphinx to access the links in the documentation output and validate whether they are working.

Reviewer 2: Optional:
```suggestion
The Starter Pack checks any links in your documentation with the Sphinx ``linkcheck`` builder.
```

Run
~~~

> Run the following command from within your documentation folder.

Reviewer 3: I've been thinking - since the word "documentation" appears in so many contexts, I'd probably process the meaning of this sentence more easily if it said something like

```suggestion
Run the following command in your `/docs` directory.
```

Though _technically_ people could name that folder something else. Discard this suggestion if you're not entirely convinced.

Reviewer 2: I flagged the same thing in [this comment](https://github.com/canonical/sphinx-docs-starter-pack/pull/573/changes#r3033845540), but my solution is more drastic.

I don't have strong opinions on the specific wording, but I think something similar could be applied here if we like that direction.

Author: With the number of checks that are now on one page, I'm expecting users to use the side nav and the links at the beginning to jump to sections which makes me hesitant to rely too much on anything established in a previous section's content. If we included it right at the beginning of the page, that *might* be enough though. 

I do agree that 'documentation' is completely overloaded and using `/docs` will help when referring to a specific directory location.

Validate links within the documentation::

   make linkcheck

Configure
~~~~~~~~~

> If you have links in the documentation that you don't want to be checked (for example, because they are local links or give random errors even though they work), you can add them to the ``linkcheck_ignore`` variable in the :file:`conf.py` file.

Reviewer 2: Optional:
```suggestion
If you have links in the documentation that you don't want to check, add them to the ``linkcheck_ignore`` variable in the :file:`conf.py` file.
```
Mainly just for concision. I'm not convinced we need this example, but I'll leave that to you.
